Userland/Libraries/LibGfx/Path.cpp

```diff
@@ -365,4 +365,166 @@ void Path::add_path(Path const& other)
+    if (pen_vertex_count % 2 == 1)
+        pen_vertex_count += 1;
+
+    Vector<FloatPoint, 128> pen_vertices;
+    pen_vertices.ensure_capacity(pen_vertex_count);
+
+    // Generate vertices for the pen (going counterclockwise). The pen does not necessarily need
+    // to be a circle (or an approximation of one), but other shapes are untested.
+    float theta = 0;
+    float theta_delta = (AK::Pi<float> * 2) / pen_vertex_count;
+    for (int i = 0; i < pen_vertex_count; i++) {
+        float sin_theta;
+        float cos_theta;
+        AK::sincos(theta, sin_theta, cos_theta);
+        pen_vertices.unchecked_append({ cos_theta * thickness / 2, sin_theta * thickness / 2 });
+        theta -= theta_delta;
+    }
+
+    auto wrapping_index = [](auto& vertices, auto index) {
+        return vertices[(index + vertices.size()) % vertices.size()];
+    };
+
+    auto angle_between = [](auto p1, auto p2) {
+        auto delta = p2 - p1;
+        return atan2f(delta.y(), delta.x());
+    };
+
+    struct ActiveRange {
+        float start;
+        float end;
+
+        bool in_range(float angle) const
+        {
+            // Note: Since active ranges go counterclockwise start > end unless we wrap around at 180 degrees
+            return ((angle <= start && angle >= end)
+                || (start <= end && angle <= start)
+                || (start <= end && angle >= end));
+        }
+    };
+
+    Vector<ActiveRange, 128> active_ranges;
+    active_ranges.ensure_capacity(pen_vertices.size());
+    for (auto i = 0; i < pen_vertex_count; i++) {
+        active_ranges.unchecked_append({ angle_between(wrapping_index(pen_vertices, i - 1), pen_vertices[i]),
+            angle_between(pen_vertices[i], wrapping_index(pen_vertices, i + 1)) });
+    }
+
+    auto clockwise = [](float current_angle, float target_angle) {
+        if (target_angle < 0)
+            target_angle += AK::Pi<float> * 2;
+        if (current_angle < 0)
+            current_angle += AK::Pi<float> * 2;
+        if (target_angle < current_angle)
+            target_angle += AK::Pi<float> * 2;
+        return (target_angle - current_angle) <= AK::Pi<float>;
+    };
+
+    Path convolution;
+    for (auto& segment : segments) {
+        RoundTrip<FloatPoint> shape { segment };
+
+        bool first = true;
+        auto add_vertex = [&](auto v) {
+            if (first) {
+                convolution.move_to(v);
+                first = false;
+            } else {
+                convolution.line_to(v);
+            }
+        };
+
+        auto shape_idx = 0u;
+
+        auto slope = [&] {
+            return angle_between(shape[shape_idx], shape[shape_idx + 1]);
+        };
+
+        auto start_slope = slope();
+        auto active = 0u;
+        for (; active < active_ranges.size(); active++) {
+            if (active_ranges[active].in_range(start_slope)) {
+                break;
+            }
+        }
```

Comment:
If all you need is the index, maybe `Vector::find_first_index_if` can help here?

Reply:
Thanks! That's handy :)